engine/src/main/java/dgca/verifier/app/engine/DefaultCertLogicEngine.kt

```diff
@@ -67,48 +67,66 @@ class DefaultCertLogicEngine(
         externalParameter: ExternalParameter,
         payload: String
     ): List<ValidationResult> {
-        return if (rules.isNotEmpty()) {
-            val validationResults = mutableListOf<ValidationResult>()
-            val dataJsonNode = prepareData(externalParameter, payload)
-            val hcertVersion = hcertVersionString.toVersion()
-            rules.forEach { rule ->
-                val ruleEngineVersion = rule.engineVersion.toVersion()
-                val schemaVersion = rule.schemaVersion.toVersion()
-                val res = when {
-                    rule.engine == CERTLOGIC_KEY && ruleEngineVersion != null && CERTLOGIC_VERSION.isGreaterOrEqualThan(
-                        ruleEngineVersion
-                    ) && hcertVersion != null && schemaVersion != null && hcertVersion.first == schemaVersion.first && hcertVersion.isGreaterOrEqualThan(
-                        schemaVersion
-                    ) ->
-                        when (jsonLogicValidator.isDataValid(
-                            rule.logic,
-                            dataJsonNode
-                        )) {
-                            true -> Result.PASSED
-                            false -> Result.FAIL
-                            else -> Result.OPEN
-                        }
-                    else -> Result.OPEN
+        if (rules.isEmpty()) return emptyList()
+
+        val dataJsonNode = prepareData(externalParameter, payload)
+        val hcertVersion: Triple<Int, Int, Int>? = hcertVersionString.toVersion()
+
+        return rules.map {
+            checkRule(
+                rule = it,
+                dataJsonNode = dataJsonNode,
+                hcertVersion = hcertVersion,
+                certificateType = certificateType
+            )
+        }
```

Comment:
Early return to prervent unnecessary nesting, then treat this as data preperation step.